Changes to the general code-base

ilevn opened the pull request:
This PR refactors a lot of of the existing code-base. Some changes may be too functional (@Fuyukai ) but I'll leave that up to the reviewers.

ilevn committed:
Replace conditional with `takeIf`

ilevn committed:
Use `buildString` instead of a manual concat.

ilevn committed:
Favour `let` over imperative assignments

ilevn committed:
Replace null conditional with `let`.

ilevn committed:
Refactor `getFinalArgs`.

ilevn committed:
Reformat KyCodeObject using `with` and `buildString`.

ilevn committed:
Rewrite `fromAnyMap` to reduce overhead.

ilevn committed:
Use `buildString`; Rename `getFrame` to `createFrame` and
adjust docstring.

ilevn committed:
Remove `hasDisassembly` and `hasStack`; Convert functions to properties.

ilevn committed:
Refactor codebase. This includes:
- Removal of explicit `this`.
- Function renames.
- Modifier changes.
- Frequent use of Kotlin receiver/extension functions

ilevn committed:
Refactor remaining codebase. This includes:
- Removal of explicit `this`.
- Function renames.
- Modifier changes.
- Frequent use of Kotlin receiver/extension functions

Fuyukai commented on src/main/kotlin/interpreter/functions/PyFunction.kt:
```diff
@@ -29,9 +29,6 @@ import green.sailor.kython.interpreter.pyobject.PyType
-            return this
-        }
-        return PyMethod(this, parent)
+        return takeIf { parent is PyNone } ?: PyMethod(this, parent)
```
Extremely against this

ilevn replied:
Why?

Fuyukai replied:
I like regular if statements over a functional if statement, especially when if is an expression, this feels like functional for the sake of being functional

ilevn replied:
Sure, I'll change it

Fuyukai commented on src/main/kotlin/interpreter/kyobject/KyCodeObject.kt:
```diff
@@ -137,27 +139,25 @@ class KyCodeObject(original: KycCodeObject) {
+        val padSize = ceil(log(instructions.size.toDouble(), 10.0)).toInt()
+
+        return buildString {
+            instructions.forEachIndexed { index, instruction ->
```
`index` args should always be `idx`

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyDict.kt:
```diff
@@ -31,27 +31,20 @@ class PyDict(val items: LinkedHashMap<out PyObject, out PyObject>) : PyObject(Py
+                    it.key as? PyObject ?: wrapPrimitive(it.key)
+                }, {
+                    it.value as? PyObject ?: wrapPrimitive(it.value)
+                })
```
I prefer my solution, this is just kinda ugly to read - plus I think the perf impact of creating Pair() will be mitigated be it by JIT magic or Valhalla in the future.

Fuyukai commented on src/main/kotlin/interpreter/pyobject/PyException.kt:
```diff
@@ -39,9 +39,9 @@ abstract class PyException(val args: PyTuple) : PyObject() {
          * Internal method for interpreter Python errors.
          */
-        fun makeWithMessage(message: String): PyException {
+        private fun makeWithMessage(message: String): PyException {
```
Shouldn't be private

Fuyukai commented on src/main/kotlin/interpreter/pyobject/types/PyBytesType.kt:
```diff
@@ -23,12 +23,10 @@ import green.sailor.kython.interpreter.iface.ArgType
     override fun newInstance(kwargs: Map<String, PyObject>): PyObject {
-        val value = kwargs["value"] ?: error("Built-in signature mismatch")
-        when (value) {
+        when (val value = kwargs["value"] ?: error("Built-in signature mismatch")) {
```
Not a fan of this, keep the temporary variable separate - plus having an `?: error(...)` in this block looks really weird.

Fuyukai commented on src/main/kotlin/interpreter/pyobject/types/PyIntType.kt:
```diff
@@ -32,8 +32,7 @@ import green.sailor.kython.interpreter.throwKy
     override fun newInstance(kwargs: Map<String, PyObject>): PyObject {
-        val value = kwargs["value"] ?: error("Built-in signature mismatch")
-        when (value) {
+        when (val value = kwargs["value"] ?: error("Built-in signature mismatch")) {
```
Same issue here

Fuyukai commented on src/main/kotlin/interpreter/pyobject/types/PyTupleType.kt:
```diff
@@ -27,10 +27,7 @@ import green.sailor.kython.interpreter.pyobject.PyType
  * Represents the type of a tuple (the `tuple` builtin).
  */
 object PyTupleType : PyType("tuple") {
-    override fun newInstance(args: Map<String, PyObject>): PyObject {
```
For consistency newInstance should always be a full function, plus this will eventually be fleshed out into one with iterators.

Fuyukai commented on src/main/kotlin/interpreter/stack/StackFrameInfo.kt:
```diff
@@ -57,30 +49,32 @@ abstract class StackFrameInfo {
-                    "    $sourceLine"
-        }
+        override val disassembly: String
+            get() {
```
gets like these can all be one line: `get() = frame.function.code.getDisassembly(frame)`

Fuyukai commented on src/main/kotlin/interpreter/stack/StackFrameInfo.kt:
```diff
@@ -90,8 +84,9 @@ abstract class StackFrameInfo {
-            return "File <builtin>, in ${frame.builtinFunction.name}"
-        }
+        override val tracebackString: String
+            get() {
```
One line

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -66,78 +66,80 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
-        return this.function.code.getLineNumber(this.bytecodePointer)
-    }
+    val lineNo: Int
+        get() {
```
One line

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -147,9 +149,7 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
-    fun returnValue(arg: Byte): PyObject {
-        return stack.pop()
-    }
+    fun returnValue(arg: Byte): PyObject = stack.pop()
```
All instructions should be a full function

Fuyukai replied:
Still needs resolving

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -158,62 +158,51 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
      * STORE_(NAME|FAST).
      */
     fun store(pool: LoadPool, arg: Byte) {
-        val idx = arg.toInt()
```
Keep these temp variables, it makes more sense to have something called idx than just passing the argval

Fuyukai replied:
Still needs resolving

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -222,37 +211,36 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
     fun callFunction(opval: Byte) {
         // CALL_FUNCTION(argc)
         // pops (argc) arguments off the stack (right to left) then invokes a function.
-        val args = opval.toInt()
```
Keep this

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -222,37 +211,36 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
-        function.module = this.function.module
-        this.stack.push(function)
-        this.bytecodePointer += 1
+        function.module = function.module
```
This needs the this, or a variable rename, this doesn't actually do anything.

Fuyukai commented on src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt:
```diff
@@ -348,9 +336,8 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
                 val second = stack.pop() as PyInt
                 stack.push(PyInt(first.wrappedInt + second.wrappedInt))
             }
-            else -> TODO("Unsupported binary op $type")
```
Keep these empty else branches for future-proofing.

Fuyukai commented on src/main/kotlin/kyc/UnKyc.kt:
```diff
@@ -99,30 +99,24 @@ open class UnKyc(protected val buf: ByteBuffer) {
-    fun readInt(): KycInt {
-        return KycInt(buf.int)
-    }
+    private fun readInt(): KycInt = KycInt(buf.int)
```
None of these should be private

Fuyukai commented on src/main/kotlin/kyc/UnKyc.kt:
```diff
@@ -174,28 +168,23 @@ open class UnKyc(protected val buf: ByteBuffer) {
-        for (x in 0 until size) {
-            map[this.readObject()] = this.readObject()
-        }
+    private fun readDict(): KycDict {
```
I prefer the imperative version, it reads easier

Fuyukai commented on src/main/kotlin/util/CPythonCompiler.kt:
```diff
@@ -28,24 +28,25 @@ import java.nio.file.Path
-            kycPyPath.toAbsolutePath().toString(),
-            path.toAbsolutePath().toString()
-        ))
+        builder.command(
```
Was this a linter thing?

ilevn replied:
Yes

ilevn committed:
Improve `fromAnyMap`

ilevn committed:
Address review requests.

ilevn committed:
Converter func getter into expression body getter

ilevn committed:
Turn `let` return into imperative return

ilevn committed:
Refine gradle `spotless` tasks and give them their own group

Fuyukai commented on src/main/kotlin/interpreter/functions/PyUserFunction.kt:
```diff
@@ -65,27 +65,17 @@ class PyUserFunction(codeObject: KyCodeObject) : PyFunction(PyUserFunctionType)
-
-        Exceptions.NAME_ERROR("Name $name is not defined").throwKy()
+        return module.attribs[name] ?: Builtins.BUILTINS_MAP[name]
+        ?: Exceptions.NAME_ERROR("Name $name is not defined").throwKy()
```
This would be more nicely formatted as:
```kotlin
        return module.attribs[name]
            ?: Builtins.BUILTINS_MAP[name]
            ?: Exceptions.NAME_ERROR("Name $name is not defined").throwKy()
```

ilevn committed:
Address more review points:
- Turn getter func body into expression body
- Don't inline when value
- Use block body for newInstance functions

Fuyukai commented on src/main/kotlin/interpreter/kyobject/KyCodeObject.kt:
```diff
@@ -122,9 +122,9 @@ class KyCodeObject(original: KycCodeObject) {
+        val buf = ByteBuffer.wrap(rawBytecode)
+
         while (buf.hasRemaining()) {
-            // prevents opcodes >128 from turning into -opcode
```
Not sure why this comment was removed

ilevn replied:
Bad C/P on my end

Fuyukai reviewed (changes requested):
Just a few nitpicks left

ilevn committed:
Address final review

ilevn committed:
Address final review
